# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [3]:
import sqlite3 as sql
import pandas as pd

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [4]:
# Create a database connection
conn = sql.connect('go_sales.sqlite')
print("Opened database successfully")

Opened database successfully


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [13]:
# Read the tables from the database
product = pd.read_sql("SELECT * FROM product", conn)
product_type = pd.read_sql("SELECT * FROM product_type", conn)
product_line = pd.read_sql("SELECT * FROM product_line", conn)
sales_staff = pd.read_sql("SELECT * FROM sales_staff", conn)
sales_branch = pd.read_sql("SELECT * FROM sales_branch", conn)
retailer_site = pd.read_sql("SELECT * FROM retailer_site", conn)
country = pd.read_sql("SELECT * FROM country", conn)
order_header = pd.read_sql("SELECT * FROM order_header", conn)
order_details = pd.read_sql("SELECT * FROM order_details", conn)
returned_item = pd.read_sql("SELECT * FROM returned_item", conn)
return_reason = pd.read_sql("SELECT * FROM return_reason", conn)
target = pd.read_sql("SELECT * FROM SALES_TARGETData", conn)
product

PRODUCT_NUMBER INTRODUCTION_DATE PRODUCT_TYPE_CODE PRODUCTION_COST MARGIN  \
0                1         15-2-2011                 1               4    .33   
1               10         15-2-2011                 1              10     .4   
2              100         15-2-2011                17               3     .5   
3              101        15-12-2019                18          305.54    .43   
4              102        10-12-2019                18          380.95    .51   
..             ...               ...               ...             ...    ...   
110             95         15-2-2011                16               3     .5   
111             96         15-2-2011                17           16.43    .28   
112             97          5-3-2013                17              25    .28   
113             98         15-2-2011                17               3     .5   
114             99         15-2-2011                17               2     .6   

       PRODUCT_IMAGE LANGUAGE              PRODUCT_NAME  \
0      P01CE1CG1.jpg       EN       TrailChef Water Bag   
1      P10CE1CG1.jpg       EN        TrailChef Utensils   
2    P100OP4FA17.jpg       EN        Insect Bite Relief   
3    P101GE5IR18.jpg       EN     Hailstorm Steel Irons   
4    P102GE5IR18.jpg       EN  Hailstorm Titanium Irons   
..               ...      ...                       ...   
110   P91OP4SS16.jpg       EN                Sun Shield   
111   P96OP4FA17.jpg       EN        Compact Relief Kit   
112   P96OP4FA17.jpg       EN  Deluxe Family Relief Kit   
113   P98OP4FA17.jpg       EN           Calamine Relief   
114   P99OP4FA17.jpg       EN               Aloe Relief   

                                           DESCRIPTION TRIAL888  
0    Lightweight, collapsible bag to carry liquids ...        T  
1    Spoon, fork and knife set made of a light yet ...        T  
2    The Insect Bite Relief helps the itching and s...        T  
3    Iron is 17-4 stainless steel.  Shafts are grap...        T  
4    Made entirely of pure titanium. The ultimate i...        T  
..                                                 ...      ...  
110  PABA free sunscreen, SPF 30, poison oak and iv...        T  
111  A personal first aid kit is recommended for ev...        T  
112  A complete medical kit suitable for families w...        T  
113  Use the Calamine Relief for allergic skin reac...        T  
114  Perfect for minor burns and sunburn, the aloe ...        T  

[115 rows x 10 columns]

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [18]:
# Convert 'PRODUCTION_COST' column to numeric
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'], errors='coerce')

# Filter the products where the production costs are between $50 and $100
filtered_product = product[(product['PRODUCTION_COST'] > 50) & (product['PRODUCTION_COST'] < 100)]

# Select only the columns for product name and production costs
resultaat = filtered_product[['PRODUCT_NAME', 'PRODUCTION_COST']]

# Show the resulting dataframe
resultaat

PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [20]:
# Convert 'MARGIN' column to numeric
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

# Filter the products where the margin is lower than 20% or higher than 60%
filtered_product = product[(product['MARGIN'] < 20) | (product['MARGIN'] > 60)]

# Select only the columns for product name and margin
resultaat = filtered_product[['PRODUCT_NAME', 'MARGIN']]

# Show the resulting dataframe
resultaat

PRODUCT_NAME  MARGIN
0         TrailChef Water Bag    0.33
1          TrailChef Utensils    0.40
2          Insect Bite Relief    0.50
3       Hailstorm Steel Irons    0.43
4    Hailstorm Titanium Irons    0.51
..                        ...     ...
110                Sun Shield    0.50
111        Compact Relief Kit    0.28
112  Deluxe Family Relief Kit    0.28
113           Calamine Relief    0.50
114               Aloe Relief    0.60

[115 rows x 2 columns]

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [38]:
# Filter the countries where payments are made in francs
franc_countries = country[country['CURRENCY_NAME'] == 'francs'].copy()


# Sort the resulting dataframe by the country column
franc_countries.sort_values('COUNTRY', inplace=True)

# Show the resulting dataframe with only the country column
franc_countries[['COUNTRY']]

COUNTRY
7       Belgium
0        France
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [39]:
# Filter the products where the margin is greater than 50%
filtered_product = product[product['MARGIN'] > 50]

# Select only the column for introduction date
resultaat = filtered_product[['INTRODUCTION_DATE']]

# Show the resulting dataframe
resultaat

Empty DataFrame
Columns: [INTRODUCTION_DATE]
Index: []

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 